aws/resource_aws_workspace.go

```diff
@@ -0,0 +1,405 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/workspaces"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/validation"
+	"github.com/terraform-providers/terraform-provider-aws/aws/internal/keyvaluetags"
+)
+
+func resourceAwsWorkspace() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsWorkspaceCreate,
+		Read:   resourceAwsWorkspaceRead,
+		Update: resourceAwsWorkspaceUpdate,
+		Delete: resourceAwsWorkspaceDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+		Schema: map[string]*schema.Schema{
+			"bundle_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"directory_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"ip_address": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"computer_name": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"state": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"root_volume_encryption_enabled": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				ForceNew: true,
+				Default:  false,
+			},
+			"user_name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"user_volume_encryption_enabled": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				ForceNew: true,
+				Default:  false,
+			},
+			"volume_encryption_key": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+			"workspace_properties": {
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"compute_type_name": {
+							Type:     schema.TypeString,
+							Optional: true,
+							Default:  workspaces.ComputeValue,
+							ValidateFunc: validation.StringInSlice([]string{
+								workspaces.ComputeValue,
+								workspaces.ComputeStandard,
+								workspaces.ComputePerformance,
+								workspaces.ComputePower,
+								workspaces.ComputePowerpro,
+								workspaces.ComputeGraphics,
+								workspaces.ComputeGraphicspro,
+							}, false),
+						},
+						"root_volume_size_gib": {
+							Type:         schema.TypeInt,
+							Optional:     true,
+							Default:      80,
+							ValidateFunc: validation.IntInSlice([]int{80, 175}),
```

Comment:
I don't' remember exactly why, but I failed to accomplish this. Will make another attempt and share the feedback. Thanks.

Reply:
@gdavison Ah, I remembered, how can I refer to attribute A in the validation func of attribute B?

Reply:
@gdavison The question about validation func for dependent attributes is still actual for me. Please could you answer the questions?

Reply:
For the `ValidateFunc`s in the resource schema, you can use `validation.Any()` to chain together validation functions.

For `root_volume_size_gib`, this could be

```
ValidateFunc: validation.Any(
    validation.IntInSlice([]int{80}),
    validation.IntBetween(175, 2000),
)
```

For final validation that the two fields are compatible, we'll have to let the AWS API handle it.

Reply:
Done. I've added one acceptance for user volume size validation, but I doubt it's the right way to test all possible cases. Please could you refer me to the good validation unit test example?